src/main/java/io/confluent/kafkarest/v2/KafkaConsumerReadTask.java

```diff
@@ -0,0 +1,232 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.kafkarest.v2;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Vector;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import io.confluent.kafkarest.entities.ConsumerRecord;
+import io.confluent.rest.exceptions.RestException;
+import io.confluent.kafkarest.KafkaRestConfig;
+import io.confluent.kafkarest.ConsumerRecordAndSize;
+import io.confluent.kafkarest.ConsumerWorkerReadCallback;
+
+import kafka.consumer.ConsumerTimeoutException;
+
+/**
+ * State for tracking the progress of a single consumer read request.
+ *
+ * To support embedded formats that require translation between the format deserialized by the Kafka
+ * decoder and the format returned in the ConsumerRecord entity sent back to the client, this class
+ * uses two pairs of key-value generic type parameters: KafkaK/KafkaV is the format returned by the
+ * Kafka consumer's decoder/deserializer, ClientK/ClientV is the format returned to the client in
+ * the HTTP response. In some cases these may be identical.
+ */
+class KafkaConsumerReadTask<KafkaK, KafkaV, ClientK, ClientV>
+    implements Future<List<ConsumerRecord<ClientK, ClientV>>> {
+
+  private static final Logger log = LoggerFactory.getLogger(KafkaConsumerReadTask.class);
+
+  private KafkaConsumerState parent;
+  private final long pollTimeout;
+  private final long maxResponseBytes;
+  private final ConsumerWorkerReadCallback<ClientK, ClientV> callback;
+  private CountDownLatch finished;
+
+  private KafkaConsumerTopicState topicState;
+  private Iterator<org.apache.kafka.clients.consumer.ConsumerRecord<ClientK, ClientV>> iter;
+  private List<ConsumerRecord<ClientK, ClientV>> messages;
+  private long bytesConsumed = 0;
+  private final long started;
+
+  // Expiration if this task is waiting, considering both the expiration of the whole task and
+  // a single backoff, if one is in progress
+  long waitExpiration;
+
+  public KafkaConsumerReadTask(KafkaConsumerState parent, String topic, long timeout, long maxBytes,
+                          ConsumerWorkerReadCallback<ClientK, ClientV> callback) {
+    this.parent = parent;
+    this.maxResponseBytes = Math.min(
+        maxBytes,
+        parent.getConfig().getLong(KafkaRestConfig.CONSUMER_REQUEST_MAX_BYTES_CONFIG));
+    this.pollTimeout = timeout <= 0 ? 5000 : timeout;
+    this.callback = callback;
+    this.finished = new CountDownLatch(1);
+
+    started = parent.getConfig().getTime().milliseconds();
+    try {
+	topicState = parent.getOrCreateTopicState(topic, pollTimeout);
```

Comment:
Doing a poll inside the KafkaConsumerState.getOrCreateTopicState is not ideal here, it should be done in doPartialRead(), with some kinda strategy that check if timeout or max_bytes limit has it and finishes the execution. In a case of failed task, the messages from failed task are used first, and which inturn could mean whole batch of messages would be lost because of no internal queue.